README.md
# SPSP: an R Package for Selecting the relevant predictors by Partitioning the Solution Paths of the Penalized Likelihood Approach

<!-- badges: start -->

[![](https://cran.r-project.org/package=SPSP)](https://img.shields.io/cran/v/SPSP?logo=R)
[![CRAN checks](https://cranchecks.info/badges/summary/SPSP)](https://cran.r-project.org/web/checks/check_results_SPSP.html)
[![](https://cranlogs.r-pkg.org/badges/grand-total/SPSP?color=blue)](https://cranlogs.r-pkg.org/badges/grand-total/SPSP)
[![](https://cranlogs.r-pkg.org/badges/last-month/SPSP?color=green)](https://cranlogs.r-pkg.org/badges/last-month/SPSP?color=green)
[![](https://cranlogs.r-pkg.org/badges/last-week/SPSP?color=yellow)](https://cranlogs.r-pkg.org/badges/last-week/SPSP?color=yellow)

<!-- badges: end -->

Overview
--------

An implementation of the feature Selection procedure by Partitioning the entire Solution Paths
(namely SPSP) to identify the relevant features rather than using a single tuning parameter. 
By utilizing the entire solution paths, this procedure can obtain better selection accuracy than 
the commonly used approach of selecting only one tuning parameter based on existing criteria, 
cross-validation (CV), generalized CV, AIC, BIC, and EBIC (Liu, Y., & Wang, P. (2018) 
https://doi.org/10.1214/18-EJS1434). It is more stable and accurate (low false positive and false negative
rates) than other variable selection approaches. In addition, it can be flexibly coupled with 
the solution paths of Lasso, adaptive Lasso, ridge regression,  and other penalized estimators.

## Installation

The `SPSP` package is currently available on [SPSP CRAN](https://CRAN.R-project.org/package=SPSP).

### Install `SPSP` development version from GitHub (recommended)

``` r
# Install the development version from GitHub
if (!requireNamespace("devtools")) install.packages("devtools")
devtools::install_github("XiaoruiZhu/SPSP")
```

### Install `SPSP` from the CRAN

``` r
# Install from CRAN
install.packages("SPSP")
```


## Example

The following example shows the R code for 

``` r
library(SPSP)
data(HihgDim)
library(glmnet)

# Use the user-friendly function SPSP() to conduct the selection by Partitioning the 
# Solution Paths (the SPSP algorithm). The user only needs to specify the independent 
# variables matrix, response, family, and \code{fitfun.SP = lasso.glmnet}. 

x <- as.matrix(HighDim[,-1])
y <- HighDim[,1]

spsp_lasso_1 <- SPSP(x = x, y = y, family = "gaussian", fitfun.SP = lasso.glmnet,
                     init = 1, standardize = FALSE, intercept = FALSE)

head(spsp_lasso_1$nonzero)
head(spsp_lasso_1$beta_SPSP)

spsp_adalasso_5 <- SPSP(x = x, y = y, family = "gaussian", fitfun.SP = adalasso.glmnet,
                        init = 5, standardize = T, intercept = FALSE)
                              
head(spsp_adalasso_5$nonzero)
head(spsp_adalasso_5$beta_SPSP)

# Use the function SPSP_step() to select the relevant predictors by partitioning the 
# solution paths based on the user provided solution paths \code{BETA}. 

lasso_fit <- glmnet(x = x, y = y, alpha = 1, intercept = FALSE)

# SPSP+Lasso method
K <- dim(lasso_fit$beta)[2]
LBETA <- as.matrix(lasso_fit$beta)
spsp_lasso_1 <- SPSP_step(x = x, y = y, BETA = LBETA, 
                          init = 1, standardize = FALSE, intercept = FALSE)

head(spsp_lasso_1$nonzero)
head(spsp_lasso_1$beta_SPSP)
```

References
----------

Liu, Y., & Wang, P. (2018). Selection by partitioning the solution paths. *Electronic Journal of Statistics*, 12(1), 1988-2017. <10.1214/18-EJS1434>



NEWS.md
# SPSP 0.1.1

## Major changes

1. Fix the function corresponding to the `intercept `argument. The intercept argument controls whether to include the intercept, but since the SPSP algorithm estimates the intercept separately. It needs to be dealt with separately.

2. The `standardize` only control the standardization of covariates.

# SPSP 0.1.0

## Major changes

1. Implement the Selection by Partitioning the Solution Paths procedure in the SPSP() and SPSP_step() functions;

2. Add fitfun.SP functions to obtain the solution paths for the SPSP algorithm;

3. Include one high-dimensional data for illustration purpose.





tests/testthat/test-SPSP.R
test_that("Check SPSP_step() use data(HighDim)", {
  skip_if_not_installed("glmnet")
  
  library(glmnet)
  
  # Use the high dimensional dataset (data(HighDim)) to test SPSP+Lasso:
  data(HighDim)  
  
  x <- as.matrix(HighDim[,-1])
  y <- HighDim[,1]
  
  lasso_fit <- glmnet(x = x, y = y, alpha = 1, intercept = FALSE)
  
  # library(plotmo) # for plot_glmnet
  # plot_glmnet(lasso_fit, label=15, xvar ="lambda")
  # coef(lasso_fit, s = 0.5)
  
  test_lm_3 <- lm(Y ~ X1 + X2 + X3 + 0, data = HighDim)
  
  # When nonzero == 0, run the following to obtain intercept
  # test_lm_int <- glm.fit(y = y, x = rep(1, length(y)), intercept = TRUE, family = gaussian())$coefficients
  
  # When nonzero >= 1, run the following to obtain intercept
  # test_lm_int <- glm.fit(y = y, x = cbind(1, x[,1:3]), intercept = FALSE, family = gaussian())$coefficients
  test_lm_int <- glm(y ~ X1 + X2 + X3 + 1, data = HighDim, family = gaussian())$coefficients
  
  
  # summary(test_lm_3)
  
  # SPSP+Lasso method
  K <- dim(lasso_fit$beta)[2]
  LBETA <- as.matrix(lasso_fit$beta)
  
  r_spsp1 <- SPSP::SPSP_step(x = x, y = y, BETA = LBETA, init = 1, standardize = FALSE, intercept = FALSE)
  est_beta_1 <- r_spsp1$beta_SPSP
  # colnames(X)[which(est_beta_1 != 0)]
  head(est_beta_1)
  
  r_spsp5 <- SPSP::SPSP_step(x = x, y = y, BETA = LBETA, init = 5, standardize = FALSE)
  est_beta_5 <- r_spsp5$beta_SPSP
  # head(est_beta_5)
  
  r_spsp5_std <- SPSP::SPSP_step(x = x, y = y, BETA = LBETA, init = 5, standardize = T, intercept = T)
  est_beta_5_std <- r_spsp5_std$beta_SPSP
  # head(est_beta_5_std)
  # r_spsp5_std$intercept
  
  # Comparison
  expect_equal(coef(test_lm_3), est_beta_1[1:3])
  expect_equal(coef(test_lm_3), est_beta_5[1:3])
  # expect_equal(coef(test_lm_3), est_beta_5_std[1:3])
  
  expect_equal(test_lm_int[1], r_spsp5_std$intercept)
})

test_that("Check SPSP() with lasso.glmnet() use data(HighDim)", {
  skip_if_not_installed("glmnet")
  
  library(glmnet)
  
  # Use the high dimensional dataset (data(HighDim)) to test SPSP+Lasso:
  data(HighDim)  
  
  x <- as.matrix(HighDim[,-1])
  y <- HighDim[,1]
  
  xstd <- scale(HighDim[,-1], center = TRUE, scale = TRUE)
  
  HighDim2 <- data.frame(Y = y, xstd)
  
  test_lm_3 <- lm(Y ~ X1 + X2 + X3 + 0, data = HighDim)
  
  test_lm_std_3 <- lm(Y ~ X1 + X2 + X3 + 0, data = HighDim2)
  
  # When nonzero == 0, run the following to obtain intercept
  # test_lm_int <- glm.fit(y = y, x = rep(1, length(y)), intercept = TRUE, family = gaussian())$coefficients
  
  # When nonzero >= 1, run the following to obtain intercept
  # test_lm_int <- glm.fit(y = y, x = cbind(1, x[,1:3]), intercept = FALSE, family = gaussian())$coefficients
  test_lm_int <- glm(y ~ X1 + X2 + X3 + 1, data = HighDim, family = gaussian())$coefficients
  
  
  lasso_fit <- glmnet(x = x, y = y, alpha = 1, intercept = FALSE)
  # coef(lasso_fit)
  
  r_spsp1 <- SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = lasso.glmnet,
                        init = 1, standardize = FALSE, intercept = FALSE)
  
  est_beta_1 <- r_spsp1$beta_SPSP
  # colnames(X)[which(est_beta_1 != 0)]
  # head(est_beta_1)
  
  r_spsp5 <- SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = lasso.glmnet,
                        init = 5, standardize = FALSE, intercept = FALSE)
  est_beta_5 <- r_spsp5$beta_SPSP
  # head(est_beta_5)
  
  r_spsp5_std <- SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = adalasso.glmnet,
                            init = 5, standardize = TRUE, intercept = FALSE)
  est_beta_5_std <- r_spsp5_std$beta_SPSP
  # head(est_beta_5_std)
  
  r_spsp5_std_int <- SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = adalasso.glmnet,
                            init = 5, standardize = TRUE, intercept = TRUE)
  est_beta_5_std_int <- r_spsp5_std_int$beta_SPSP
  # head(est_beta_5_std_int)
  # r_spsp5_std_int$intercept
  
  # Check if coefficients are same as ols
  expect_equal(coef(test_lm_3), est_beta_1[1:3])
  expect_equal(coef(test_lm_3), est_beta_5[1:3])
  expect_equal(coef(test_lm_std_3), est_beta_5_std[1:3])
  expect_equal(coef(test_lm_std_3), est_beta_5_std_int[1:3])
  
  # Check if intercept is correct
  expect_true(is.na(r_spsp1$intercept))
  expect_true(is.na(r_spsp5$intercept))
  expect_true(is.na(r_spsp5_std$intercept))
  expect_true(!is.na(r_spsp5_std_int$intercept))
})

test_that("Check SPSP() with use data(Boston)", {
  skip_if_not_installed("MASS")
  skip_if_not_installed("glmnet")
  
  library(glmnet)
  library(MASS)
  # Boston Housing data from http://archive.ics.uci.edu/ml/datasets/Housing
  data(Boston, package="MASS")
  colnames(Boston) 
  summary(Boston)
  
  Boston2 <- subset(Boston, crim<=3.2)
  # Boston2 <- Boston
  # dim(Boston2)
  x <- as.matrix(Boston2[,-14]); y <- Boston2[,14] 
  summary(x)
  x[,"crim"] <- log(x[,"crim"])
  x[,"tax"] <- log(x[,"tax"])
  x[,"lstat"] <- log(x[,"lstat"])
  x[,"dis"] <- log(x[,"dis"])
  # x[,"rad"] <- log(x[,"rad"])
  y <- scale(y)
  x[,"age"] <- log(x[,"age"])
  
  for (i in 1:(ncol(x))){
    x[,i] <- scale(x[,i])
  }
  # summary(x)
  
  # round(cor(x), 2)
  # round(cor(Boston[,-14]), 2)
  # summary(y)
  
  expect_error(SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = lasso.glmnet,
                          init = 5, standardize = T, intercept = FALSE),
               NA)
  
  expect_error(SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = adalasso.glmnet,
                          init = 5, standardize = T, intercept = FALSE),
               NA)
  
  expect_error(SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = adalassoCV.glmnet,
                          init = 5, standardize = T, intercept = FALSE),
               NA)
  
  expect_error(SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = ridge.glmnet,
                          init = 5, standardize = T, intercept = FALSE),
               NA)
  
  err <- expect_error(SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = lasso.lars,
                          init = 5, standardize = T, intercept = FALSE))
  expect_equal(err$message, "The function lasso.lars() is currently under development.")
  
})

test_that("Check penalty.factor argument in glmnet() for SPSP()", {
  skip_if_not_installed("MASS")
  skip_if_not_installed("glmnet")
  
  library(glmnet)
  library(MASS)
  # Boston Housing data from http://archive.ics.uci.edu/ml/datasets/Housing
  data(Boston, package="MASS")
  colnames(Boston) 
  summary(Boston)
  
  Boston2 <- subset(Boston, crim<=3.2)
  # Boston2 <- Boston
  # dim(Boston2)
  x <- as.matrix(Boston2[,-14]); medv <- Boston2[,14] 
  summary(x)
  x[,"crim"] <- log(x[,"crim"])
  x[,"tax"] <- log(x[,"tax"])
  x[,"lstat"] <- log(x[,"lstat"])
  x[,"dis"] <- log(x[,"dis"])
  # x[,"rad"] <- log(x[,"rad"])
  medv <- scale(medv)
  x[,"age"] <- log(x[,"age"])
  
  for (i in 1:(ncol(x))){
    x[,i] <- scale(x[,i])
  }
  # summary(x)
  
  # round(cor(x), 2)
  # round(cor(Boston[,-14]), 2)
  # summary(y)
  
  expect_error(SPSP::SPSP(x = x, y = medv, family = "gaussian", fitfun.SP = lasso.glmnet,
                          init = 5, standardize = T, intercept = FALSE),
               NA)
  
  test1 <- SPSP::SPSP(x = x, y = medv, family = "gaussian", fitfun.SP = lasso.glmnet,
                      args.fitfun.SP = list(penalty.factor = c(rep(1, 5), 0, rep(1, 6), 0)),
                      init = 1, standardize = T, intercept = FALSE)
  # test1$nonzero

  fit1 <- attr(test1, "glmnet.fit")
  # fit1$beta
  
  # library(plotmo) 
  
  # plot_glmnet(fit1, label=15, xvar ="lambda")
  
  expect_error(SPSP::SPSP(x = x, y = medv, family = "gaussian", fitfun.SP = adalassoCV.glmnet,
                          init = 5, standardize = T, intercept = FALSE),
               NA)
  
  test2 <- SPSP::SPSP(x = x, y = medv, family = "gaussian", fitfun.SP = adalassoCV.glmnet,
                      args.fitfun.SP = list(penalty.factor = c(rep(1, 5), 0, rep(1, 6), 0)),
                      init = 1, standardize = T, intercept = FALSE)
  # test2$nonzero
  fit2 <- attr(test2, "glmnet.fit")
  
  # plot_glmnet(fit2, label=15, xvar ="lambda")
  
})


R/data.R
#' A high dimensional dataset with n equals to 200 and p equals to 500. 
#' 
#' A dataset with 200 observations and 500 dimensions is generated from the following process:
#' linear regression model with only first three non-zero coefficients equal to 3, 2, and 1.5 respectively. 
#' The covariates are correlated with AR structure (rho=0.3). The error term is normally distributed with
#' zero mean and sd equals to 0.5.
#' 
#'
#' @docType data
#'
#' @keywords datasets
#'
#' @name HighDim
#'
#' @usage
#' data(HighDim)
#'
#' @examples
#' 
#' # HighDim dataset is generated from the following process:
#' n <- 200; p <- 500; sigma <- 0.5
#' beta <- rep(0, p); nonzero <- c(1, 2, 3); zero <- setdiff(1:p, nonzero)
#' beta[nonzero] <- c(3, 2, 1.5)
#' Sigma <- 0.3^(abs(outer(1:p,1:p,"-")))
#' library(MASS)
#' X <- mvrnorm(n, rep(0,p), Sigma)
#' error <- rnorm(n, 0, sigma)
#' 
#' X <- apply(X, 2, scale) * sqrt(n)/sqrt(n-1)
#' error <- error - mean(error)
#' 
#' Y <- X %*% beta + error
#' HighDim <- data.frame(Y, X)
#' head(HighDim)
#' 
#' 
NULL


R/SPSP.R
#' Selection by Partitioning the Solution Paths
#'
#' An implementation of the feature Selection procedure by Partitioning the entire Solution Paths
#' (namely SPSP) to identify the relevant features rather than using a single tuning parameter. 
#' By utilizing the entire solution paths, this procedure can obtain better selection accuracy than 
#' the commonly used approach of selecting only one tuning parameter based on existing criteria, 
#' cross-validation (CV), generalized CV, AIC, BIC, and EBIC (Liu, Y., & Wang, P. (2018)). It is 
#' more stable and accurate (low false positive and false negative rates) than other variable 
#' selection approaches. In addition, it can be flexibly coupled with the solution paths of Lasso, 
#' adaptive Lasso, ridge regression, and other penalized estimators.
#'
#' @details This package includes two main functions and several functions (fitfun.SP) to obtains
#' the solution paths. The \code{SPSP} function allows users to specify the penalized likelihood 
#' approaches that will generate the solution paths for the SPSP procedure. Then this function 
#' will automatically partitioning the entire solution paths. Its key idea is to classify variables
#' as relevant or irrelevant at each tuning parameter and then to select all of the variables 
#' which have been classified as relevant at least once. The \code{SPSP_step} purely apply the 
#' partitioning step that needs the solution paths as the input. In addition, there are several 
#' functions to obtain the solution paths. They can be used as an input of \code{fitfun.SP} argument.
#' 
#'
#' @docType package
#'
#' @author Xiaorui (Jeremy) Zhu, \email{zhuxiaorui1989@@gmail.com}, \cr Yang Liu, \email{yliu23@@fhcrc.org}, \cr
#' Peng Wang, \email{wangp9@@ucmail.uc.edu}
#' 
#' @references Liu, Y., & Wang, P. (2018). Selection by partitioning the solution paths. 
#' \emph{Electronic Journal of Statistics}, 12(1), 1988-2017. <10.1214/18-EJS1434>
#' 
#' @name SPSP-package
#' 
#' @useDynLib SPSP
#'
NULL

## Generic implementation of SPSP
SPSP <- function(x, ...) {
  UseMethod("SPSP", x)
}

#' Selection by partitioning the solution paths of Lasso, Adaptive Lasso, and Ridge penalized regression.
#' 
#' A user-friendly function to conduct the selection by Partitioning the Solution Paths (the SPSP algorithm). The
#' user only needs to specify the independent variables matrix, response, family, and \code{fitfun.SP}. 
#'
#' @param x A matrix with all independent variables, of dimension n by p; each row is an observation vector with p variables. 
#' @param y Response variable. Quantitative for \code{family="gaussian"} or \code{family="poisson"} (non-negative counts). 
#' For \code{family="binomial"} should be either a factor with two levels.
#' 
#' @param family Response type. Either a character string representing one of the built-in families,
#' or else a glm() family object.
#' @param fitfun.SP A function to obtain the solution paths for the SPSP algorithm. This function takes the arguments 
#' x, y, family as above, and additionally the standardize and intercept and others in \code{\link[glmnet]{glmnet}} 
#' or \code{\link[lars]{lars}}. The function fit the model with lasso, adaptive lasso, or ridge regression to 
#' return the solution path of the corresponding penalized likelihood approach.
#' \describe{
#'   \item{\code{lasso.glmnet}}{lasso selection from \code{\link[glmnet]{glmnet}}.}
#'   \item{\code{adalasso.glmnet}}{adaptive lasso selection using the \code{lambda.1se} from cross-validation lasso method
#'   to obtain initial coefficients. It uses package \code{\link[glmnet]{glmnet}}.}
#'   \item{\code{adalassoCV.glmnet}}{adaptive lasso selection using the \code{lambda.1se} from cross-validation adaptive 
#'   lasso method to obtain initial coefficients. It uses package \code{\link[glmnet]{glmnet}}.}
#'   \item{\code{ridge.glmnet}}{use ridge regression to obtain the solution path.}
#'   \item{\code{lasso.lars}}{use lasso selection in \code{\link[lars]{lars}} to obtain the solution path.}
#' } 
#' @param standardize logical argument. Should conduct standardization before the estimation? Default is TRUE.
#' @param intercept logical. If x is a data.frame, this argument determines if the resulting model matrix should contain 
#' a separate intercept or not. Default is TRUE.
#' @param args.fitfun.SP A named list containing additional arguments that are passed to the fitting function;
#' see also argument \code{args.fitfun.SP} in do.call.
#' @param ... Additional optional arguments.
#'
#' @importFrom Rcpp evalCpp
#' @importFrom stats glm.fit coef 
#' @importFrom glmnet glmnet cv.glmnet
#' 
#' @return An object of class \code{"SPSP"} is a list containing at least the following components:
#' \item{\code{beta_SPSP}}{the estimated coefficients of SPSP selected model;}
#' \item{\code{S0}}{the estimated relevant sets;}
#' \item{\code{nonzero}}{the selected covariates;}
#' \item{\code{zero}}{the covariates that are not selected;}
#' \item{\code{thres}}{the boundaries for abs(beta);}
#' \item{\code{R}}{the sorted adjacent distances;}
#' \item{\code{intercept}}{the estimated intercept when \code{intercept == T}.}
#' 
#' This object has attribute contains: 
#' 
#' \item{\code{glmnet.fit}}{the fitted penalized regression within the input function \code{fitfun.SP};}
#' \item{\code{family}}{the family of fitted object;}
#' \item{\code{fitfun.SP}}{the function to obtain the solution paths for the SPSP algorithm;}
#' \item{\code{args.fitfun.SP}}{a named list containing additional arguments for the function \code{fitfun.SP}.}
#' 
#' 
#' @export
#'
#' @examples
#' data(HighDim)
#' library(glmnet)
#' # Use the high dimensional dataset (data(HighDim)) to test SPSP+Lasso and SPSP+AdaLasso:
#' data(HighDim)  
#' 
#' x <- as.matrix(HighDim[,-1])
#' y <- HighDim[,1]
#' 
#' spsp_lasso_1 <- SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = lasso.glmnet,
#'                            init = 1, standardize = FALSE, intercept = FALSE)
#' 
#' head(spsp_lasso_1$nonzero)
#' head(spsp_lasso_1$beta_SPSP)
#' 
#' spsp_adalasso_5 <- SPSP::SPSP(x = x, y = y, family = "gaussian", fitfun.SP = adalasso.glmnet,
#'                               init = 5, standardize = TRUE, intercept = FALSE)
#'                               
#' head(spsp_adalasso_5$nonzero)
#' head(spsp_adalasso_5$beta_SPSP)
#' 
#' 
#' 
SPSP <- 
  function(x, 
           y, 
           family = c("gaussian", "binomial"),
           fitfun.SP = lasso.glmnet, 
           args.fitfun.SP = list(), 
           standardize = TRUE, 
           intercept = TRUE, 
           ...) {
  
  this.call <- match.call()
  family <- match.arg(family)
  
  glmnet_T <- inherits(fitfun.SP, "glmnet")
  
  fit_mod_SP <- do.call(fitfun.SP, c(list(x = x, y = y, family = family, 
                                          standardize = standardize, intercept = intercept),
                                     args.fitfun.SP))
  
  # Conduct the SPSP step that use the above solution path to obtain relevant predictors.
  SPSP_temp <- SPSP_step(x = x, y = y, BETA = fit_mod_SP$beta, standardize = standardize, intercept = intercept, ...)

  # Assign attributes and class
  attr(SPSP_temp, "glmnet.fit") <- fit_mod_SP
  attr(SPSP_temp, "family") <- family
  attr(SPSP_temp, "fitfun.SP") <- fitfun.SP
  attr(SPSP_temp, "args.fitfun.SP") <- args.fitfun.SP
  
  class(SPSP_temp) <- c("SPSP", class(fit_mod_SP), class(SPSP_temp))
  
  return(SPSP_temp)
}


#' The selection step with the input of the solution paths.
#' 
#' A function to select the relevant predictors by partitioning the solution paths (the SPSP algorithm) 
#' based on the user provided solution paths \code{BETA}. 
#'
#' @param x independent variables as a matrix, of dimension nobs x nvars; each row is an observation vector. 
#' @param y response variable. Quantitative for \code{family="gaussian"} or \code{family="poisson"} (non-negative counts). 
#' For \code{family="binomial"} should be either a factor with two levels.
#' 
#' @param family either a character string representing one of the built-in families, or else a glm() family object. 
#' @param BETA the solution paths obtained from a prespecified fitting step \code{fitfun.SP = lasso.glmnet} etc. It must be 
#' a p by k matrix, should be thicker and thicker, each column corresponds to a lambda, and lambda gets smaller and smaller.
#' It is just the returned value \code{beta} from a \code{glmnet} object. 
#' @param standardize whether need standardization.
#' @param intercept logical. If x is a data.frame, this argument determines if the resulting model matrix should contain 
#' a separate intercept or not.
#' @param init initial coefficients, starting from init-th estimator of the solution paths. The default is 1. 
#' @param R sorted adjacent distances, default is NULL. Will be calculated inside.
#' @param ... Additional optional arguments.
#' 
#'
#' @examples 
#' data(HighDim)
#' library(glmnet)
#' 
#' x <- as.matrix(HighDim[,-1])
#' y <- HighDim[,1]
#' 
#' lasso_fit <- glmnet(x = x, y = y, alpha = 1, intercept = FALSE)
#' 
#' # SPSP+Lasso method
#' K <- dim(lasso_fit$beta)[2]
#' LBETA <- as.matrix(lasso_fit$beta)
#' 
#' spsp_lasso_1 <- SPSP_step(x = x, y = y, BETA = LBETA, 
#'                           init = 1, standardize = FALSE, intercept = FALSE)
#' head(spsp_lasso_1$nonzero)
#' head(spsp_lasso_1$beta_SPSP)
#' 
#' @return A list containing at least the following components:
#' \item{\code{beta_SPSP}}{the estimated coefficients of SPSP selected model;}
#' \item{\code{S0}}{the estimated relevant sets;}
#' \item{\code{nonzero}}{the selected covariates;}
#' \item{\code{zero}}{the covariates that are not selected;}
#' \item{\code{thres}}{the boundaries for abs(beta);}
#' \item{\code{R}}{the sorted adjacent distances;}
#' \item{\code{intercept}}{the estimated intercept when \code{intercept == T}.}
#' 
#' This object has attribute contains: 
#' 
#' \item{\code{glmnet.fit}}{the fitted penalized regression within the input function \code{fitfun.SP};}
#' \item{\code{family}}{the family of fitted object;}
#' \item{\code{fitfun.SP}}{the function to obtain the solution paths for the SPSP algorithm;}
#' \item{\code{args.fitfun.SP}}{a named list containing additional arguments for the function \code{fitfun.SP}.}
#' 
#' 
#' @export
#' 
SPSP_step <- 
  function(x, 
           y, 
           family = c("gaussian", "binomial"), 
           BETA,
           standardize=TRUE,
           intercept = TRUE, 
           init = 1, 
           R = NULL, 
           ...) {
  
  this.call <- match.call()
  family <- match.arg(family)
  
  if (is.character(family)) 
    family <- get(family, mode = "function", envir = parent.frame())
  if (is.function(family)) 
    family <- family()
  if (is.null(family$family)) {
    print(family)
    stop("'family' not recognized")
  }
  if (!is.logical(intercept)) 
    stop("Please specify intercept == TRUE/FALSE.")
  if (!is.logical(standardize)) 
    stop("Please specify stardardize == TRUE/FALSE.")
  
  
  if (standardize == TRUE) { # standardize == TRUE
    x <- scale(x, center = TRUE, scale = TRUE)
  }
  
  n <- dim(x)[1] # size
  p <- dim(x)[2] # dimension
  
  K <- dim(BETA)[2]
  BETA <- as.matrix(BETA[,K:1]) 
  # BETA for the following should be a p by k matrix;should be sparser and sparser
  # each column corresponds to a lambda, and lambda gets larger and larger
  
  K <- dim(BETA)[2] # the number of the tuning parameter lambda
  
  if (is.null(R)) {
    #### select the default the tuning parameter if it is not given
    gap0 <- sort(diff(c(0,sort(abs(BETA[,init])))), decreasing = TRUE)
    # sorted adjacent distances 
    R <- ifelse(gap0[2]!=0, as.numeric(gap0[1]/gap0[2]), as.numeric(gap0[1]))   
  }
  
  ### only consider the rows with at least one non-zero beta
  colsbeta <- colSums(BETA)
  K2 <- max(which(colsbeta!=0))
  
  thres <- c() ## boundaries for abs(beta)
  S0 <- list() # estimated relevant sets
  S0c <- list() # estimated irrelevant sets
  
  # the initial values; 
  thres_tmp <- max(abs(BETA[, 1]))
  S0_tmp <- which(abs(BETA[, 1]) > thres_tmp)
  S0c_tmp <- which(abs(BETA[, 1])  <=  thres_tmp)
  
  # loop for the update
  if (K2 >= 1) {
    for (k in 1:K2) {
      
      beta_abs <- abs(BETA[, k])
      beta_sort <- sort(beta_abs)
      
      # update the initials
      thres[k] <- ifelse(length(S0c_tmp) >= 1 ,max(beta_abs[S0c_tmp]),0)
      S0_tmp <- which(abs(BETA[, k]) > thres[k])
      S0c_tmp <- which(abs(BETA[, k]) <= thres[k])
      
      S0[[k]] <- S0_tmp
      S0c[[k]] <- S0c_tmp
      
      if (length(S0c[[k]]) >=1) {
        
        gap <- diff(c(0, beta_sort))
        
        # the distance between current relevant and irrelevant sets
        gap_10 <- ifelse(length(S0c_tmp) == p, 0, gap[length(S0c_tmp) + 1])
        
        # gap for the current irrelevant set: S0c_tmp
        gap_0 <- gap[1:length(S0c_tmp)]
        o1 <- which.max(gap_0)
        gap_01 <- max(gap_0)
        
        gap_02 <- ifelse(o1>1, max(gap[1:(o1-1)]), 0)
        
        if (gap_10  <=  R * gap_01 & gap_01 >= R * gap_02 ) {
          
          thres[k] <- ifelse(o1>1, beta_sort[o1-1], 0)
          
          S0_tmp <- which(abs(BETA[, k]) > thres[k])
          S0c_tmp <- which(abs(BETA[, k]) <= thres[k])
          
          S0[[k]] <- S0_tmp
          S0c[[k]] <- S0c_tmp
          
        }
      }
      
    }
  }
  
  index <- rep(1,p)
  
  for(i in 1:p) {
    if (all(abs(BETA[i, 1:K2]) <= thres)) index[i] <- 0
  }
  
  nz <- which(index == 1)
  z <- which(index == 0)
  beta_SPSP <- rep(0, p)
  
  if (intercept == TRUE) { # intercept == TRUE
    if (length(nz) >= 1) {
      xc <- x[, nz]
      xc1 <- cbind(1, xc)
      if (length(nz) < n) { # if p < n
        betac1 <- glm.fit(y = y, x = xc1, intercept = FALSE, family = family)$coefficients
      } else { # # if p >= n, use ridge reg to provide final coefficients
        # betac1 <- solve(t(xc1)%*%xc1+0.001*diag(length(nz)+1))%*%t(xc1)%*%y # Very slow
        # betac1 <- crossprod(solve(crossprod(xc1)+0.001*diag(length(nz)+1)), crossprod(xc1,y)) # Faster
        betac1 <- solve((crossprod(xc1) + 0.001 * diag(length(nz) + 1)), crossprod(xc1, y)) # Fastest
      }
      intercept_est <- betac1[1]
      betac <- betac1[-1]
      
      beta_SPSP[nz] <- betac  
    } else { # if no variable is selected and intercept == TRUE, run a model with intercept only
      intercept_est <- glm.fit(y = y, x = rep(1, n), intercept = FALSE, family = family)$coefficients
      betac <- NA
    } 
    
  } else { # intercept == FALSE
    intercept_est <- NA
    if (length(nz) >= 1) {
      xc <- x[, nz]
      
      if (length(nz) < n) {
        betac <- glm.fit(y = y, x = xc, intercept = FALSE, family = family)$coefficients
      } else { # change all as ridge reg to provide final coefficients
        # betac <- solve(t(xc)%*%xc+0.001*diag(length(nz)))%*%t(xc)%*%y # Very slow
        # betac <- crossprod(solve(crossprod(xc)+0.001*diag(length(nz))), crossprod(xc,y)) # Faster
        betac <- solve((crossprod(xc) + 0.001*diag(length(nz))), crossprod(xc, y)) # Fastest
      }
      
      beta_SPSP[nz] <- betac  
    }
    
    else if (!intercept) {
      stop("No variable is selected and no intercept.")
    }
  }  
  
  # Change the display and format of some results
  names(beta_SPSP) <- colnames(x)
  nz_name <- colnames(x)[nz]
  z_name <- colnames(x)[z]
  names(intercept_est) <- "(Intercept)"
  
  SPSP <- list(beta_SPSP = beta_SPSP, 
               S0 = S0,
               nonzero = nz_name, zero = z_name,
               thres = thres, R = R, 
               intercept = intercept_est)
  
  attr(SPSP, "thres") <- thres ## boundaries for abs(beta)
  attr(SPSP, "R") <- R ## sorted adjacent distances
  
  attr(SPSP, "y") <- y
  attr(SPSP, "x") <- x
  attr(SPSP, "init") <- init
  
  return(SPSP)
}


